cpp/include/raft/comms/mpi_comms.hpp

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdio>
+#include <memory>
+
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include <mpi.h>
+#include <nccl.h>
+
+#include <raft/cudart_utils.h>
+#include <raft/comms/comms.hpp>
+#include <raft/comms/util.hpp>
+#include <raft/error.hpp>
+#include <raft/handle.hpp>
+
+#define MPI_TRY(call)                                                          \
+  do {                                                                         \
+    int status = call;                                                         \
+    if (MPI_SUCCESS != status) {                                               \
+      int mpi_error_string_lenght = 0;                                         \
+      char mpi_error_string[MPI_MAX_ERROR_STRING];                             \
+      MPI_Error_string(status, mpi_error_string, &mpi_error_string_lenght);    \
+      RAFT_EXPECTS(MPI_SUCCESS == status, "ERROR: MPI call='%s'. Reason:%s\n", \
+                   #call, mpi_error_string);                                   \
+    }                                                                          \
+  } while (0)
+
+#define MPI_TRY_NO_THROW(call)                                              \
+  do {                                                                      \
+    int status = call;                                                      \
+    if (MPI_SUCCESS != status) {                                            \
+      int mpi_error_string_lenght = 0;                                      \
+      char mpi_error_string[MPI_MAX_ERROR_STRING];                          \
+      MPI_Error_string(status, mpi_error_string, &mpi_error_string_lenght); \
+      printf("MPI call='%s' at file=%s line=%d failed with %s ", #call,     \
+             __FILE__, __LINE__, mpi_error_string);                         \
+    }                                                                       \
+  } while (0)
+
+namespace raft {
+namespace comms {
+
+static MPI_Datatype get_mpi_datatype(const datatype_t datatype) {
```

Comment:
Why's this static?

And this function can be constexpr (so if datatype is available in compile time, this function will be evaluated in compile time).

Reply:
See comment: https://github.com/rapidsai/raft/pull/63#discussion_r479663183